firestore/src/main/user_data_converter_main.cc

```diff
@@ -273,43 +291,63 @@ model::FieldValue::Array UserDataConverter::ParseArray(
     ThrowInvalidData(context, "Nested arrays are not supported");
   }
 
-  model::FieldValue::Array result;
+  Message<google_firestore_v1_Value> result;
+  result->which_value_type = google_firestore_v1_Value_array_value_tag;
+  result->array_value.values_count = CheckedSize(input.size());
+  result->array_value.values = nanopb::MakeArray<google_firestore_v1_Value>(
+      result->array_value.values_count);
 
   for (size_t i = 0; i != input.size(); ++i) {
-    absl::optional<model::FieldValue> parsed =
-        ParseData(input[i], context.ChildContext(i));
+    auto parsed = ParseData(input[i], context.ChildContext(i));
     if (!parsed) {
-      parsed = model::FieldValue::Null();
+      parsed = NullValue();
     }
-    result.push_back(std::move(parsed).value());
+    result->array_value.values[i] = *parsed->release();
   }
 
-  return result;
+  return std::move(result);
 }
 
-model::ObjectValue UserDataConverter::ParseMap(const MapFieldValue& input,
-                                               ParseContext&& context) const {
+Message<google_firestore_v1_Value> UserDataConverter::ParseMap(
+    const MapFieldValue& input, ParseContext&& context) const {
+  Message<google_firestore_v1_Value> result;
+  result->which_value_type = google_firestore_v1_Value_map_value_tag;
+  result->map_value = {};
+
   if (input.empty()) {
     const model::FieldPath* path = context.path();
     if (path && !path->empty()) {
       context.AddToFieldMask(*path);
     }
-    return model::ObjectValue{};
-  }
+  } else {
+    // Compute the final size of the fields array, which contains an entry for
+    // all fields that are not FieldValue sentinels
+    pb_size_t count = 0;
+    for (const auto& kv : input) {
+      if (!IsSentinel(kv.second)) {
+        ++count;
+      }
+    }
 
-  model::FieldValue::Map result;
-  for (const auto& kv : input) {
-    const std::string& key = kv.first;
-    const FieldValue& value = kv.second;
+    result->map_value.fields_count = count;
+    result->map_value.fields =
+        nanopb::MakeArray<google_firestore_v1_MapValue_FieldsEntry>(count);
 
-    absl::optional<model::FieldValue> parsed =
-        ParseData(value, context.ChildContext(key));
-    if (parsed) {
-      result = result.insert(key, std::move(parsed).value());
+    pb_size_t index = 0;
+    for (const auto& kv : input) {
+      const std::string& key = kv.first;
+      const FieldValue& value = kv.second;
+
+      auto parsedValue = ParseData(value, context.ChildContext(key));
+      if (parsedValue) {
+        result->map_value.fields[index].key = nanopb::MakeBytesArray(key);
+        result->map_value.fields[index].value = *parsedValue->release();
+        ++index;
+      }
     }
   }
 
-  return model::ObjectValue::FromMap(std::move(result));
+  return std::move(result);
```

Comment:
It's not necessary (and sometimes can be slightly harmful to performance) to move when returning local variables.

Reply:
Fixed